common/date_handler.py

```diff
@@ -0,0 +1,72 @@
+from datetime import datetime
+from dateutil.parser import parse
+
+from common.exceptions import BadRequestError
+from common.constants import Constants
+
+
+class DateHandler:
+    """
+    Utility class to deal with dates. Currently, this class converts dates between
+    strings and `datetime.datetime` objects as well as detecting whether a string is
+    likely to be a date.
+    """
+
+    @staticmethod
+    def is_str_a_date(potential_date):
+        """
+        This function identifies if a string contains a date. This function doesn't
+        detect which format the date is, just if there's a date or not.
+
+        :param potential_date: String data that could contain a date of any format
+        :type potential_date: :class:`str`
+        :return: Boolean to signify whether `potential_date` is a date or not
+        """
+
+        try:
+            # Disabled fuzzy to avoid picking up dates in things like descriptions etc.
+            parse(potential_date, fuzzy=False)
+            return True
+        except ValueError:
+            return False
+
+    @staticmethod
+    def str_to_datetime_object(data):
+        """
+        Convert a string to a `datetime.datetime` object. This is commonly used when
+        storing user input in ICAT (using the Python ICAT backend).
+
+        Python 3.7+ has support for `datetime.fromisoformat()` which would be a more
+        elegant solution to this conversion operation since dates are converted into ISO
+        format within this file, however, the production instance of this API is
+        typically built on Python 3.6, and it doesn't seem of enough value to mandate
+        3.7 for a single line of code
+
+        :param data: Single data value from the request body
+        :type data: Data type of the data as per user's request body
+        :return: Date converted into a :class:`datetime` object
+        :raises BadRequestError: If the date is entered in the incorrect format, as per
+            `Constants.ACCEPTED_DATE_FORMAT`
+        """
+
+        try:
+            data = datetime.strptime(data, Constants.ACCEPTED_DATE_FORMAT)
+        except ValueError:
+            raise BadRequestError(
+                "Bad request made, the date entered is not in the correct format. Use"
+                f" the {Constants.ACCEPTED_DATE_FORMAT} format to submit dates to the"
+                " API"
+            )
+
+        return data
+
+    @staticmethod
+    def datetime_object_to_str(date_obj):
+        """
+        Convert a datetime object to a string so it can be outputted in JSON
+
+        :param date_obj: Datetime object from data from an ICAT entity
+        :type date_obj: :class:`datetime.datetime`
+        :return: Datetime (of type string) in the agreed format
+        """
+        return date_obj.replace(tzinfo=None).strftime(Constants.ACCEPTED_DATE_FORMAT)
```

Comment:
I can confirm that timezones are stored in ICAT and returned in the JSON results from the entitymanager endpoint. We should preserve the timezones of any datetimes we process.

Reply:
Python ICAT outputs data in the following format:
`2019-03-02 23:44:40+00:00`

This can be represented using date codes in the following way:
`"%Y-%m-%d %H:%M:%S%z"`

In an attempt to make the change to 'round-trip' datetimes using timezone data, I changed `Constants.ACCEPTED_DATE_FORMAT` to the format above and tested. I encountered a bug as described here (`%z` is for +0000, not the ISO version of timezone info which is +00:00):
https://bugs.python.org/msg169952

The end of the issue which the above message is a part of links to a GitHub diff which says that `datetime.datetime.fromisoformat` was added to Python as a fix for this:
https://bugs.python.org/msg313105

However, I've previously ruled that function out due to a lack of Python 3.6 support (https://docs.python.org/3/library/datetime.html#datetime.date.fromisoformat), which is the version used in production for the API:
https://github.com/ral-facilities/datagateway-api/blob/3702123dd976bf482b48198690ca9194cc79973c/common/date_handler.py#L39

I could apply the new accepted date format (i.e. add `%z`) and just have users use the timezone data without a colon, but I'm aware that's not the ISO format. Any suggestions? Thought this would take 5 minutes to fix 😛

Reply:
Looking at [this](https://github.com/icatproject/python-icat/blob/09ed411782de5f9e8b5739d3d75019e123093518/tests/test_01_helper.py#L143) test in the python-icat test suite, it looks like python-icat uses a [helper function](https://github.com/icatproject/python-icat/blob/09ed411782de5f9e8b5739d3d75019e123093518/icat/helper.py#L143) to [call](https://github.com/icatproject/python-icat/blob/09ed411782de5f9e8b5739d3d75019e123093518/icat/helper.py#L173) the datetime conversion provided by the underlying suds library. (Python-ICAT works on python 3.4+). Do you think you could use the python-icat helper function or call the suds function? Both will be available to you to import.

Reply:
As per e8c13e9ad184fdabb231b6647e40b964fe909a6c, I've used the helper function to convert string to datetime since it can easily differentiate between the two types of `suds` library. It's allows the conversion from ISO format but also allows for conversion of other formats so it has some flexibility. I've used the `.isoformat()` (available from at least Python 3.5) to convert from datetime to string as that puts it in the correct format. Hopefully this is resolved now :)